pkg/scheduler/sorters.go

```diff
@@ -47,39 +39,101 @@ func sortQueue(queues []*SchedulingQueue, sortType SortType) {
 			return comp < 0
 		})
 	}
+	metrics.GetSchedulerMetrics().ObserveQueueSortingLatency(sortingStart)
 }
 
-func sortApplications(apps []*SchedulingApplication, sortType SortType, globalResource *resources.Resource) {
-	// TODO add latency metric
+func sortApplications(apps map[string]*SchedulingApplication, sortType policies.SortPolicy, globalResource *resources.Resource) []*SchedulingApplication {
+	sortingStart := time.Now()
+	var sortedApps []*SchedulingApplication
 	switch sortType {
-	case FairSortPolicy:
+	case policies.FairSortPolicy:
+		sortedApps = filterOnPendingResources(apps)
 		// Sort by usage
-		sort.SliceStable(apps, func(i, j int) bool {
-			l := apps[i]
-			r := apps[j]
+		sort.SliceStable(sortedApps, func(i, j int) bool {
+			l := sortedApps[i]
+			r := sortedApps[j]
 			return resources.CompUsageRatio(l.getAssumeAllocated(), r.getAssumeAllocated(), globalResource) < 0
 		})
-	case FifoSortPolicy:
+	case policies.FifoSortPolicy:
+		sortedApps = filterOnPendingResources(apps)
 		// Sort by submission time oldest first
-		sort.SliceStable(apps, func(i, j int) bool {
-			l := apps[i]
-			r := apps[j]
+		sort.SliceStable(sortedApps, func(i, j int) bool {
+			l := sortedApps[i]
+			r := sortedApps[j]
 			return l.ApplicationInfo.SubmissionTime < r.ApplicationInfo.SubmissionTime
 		})
+	case policies.StateAwarePolicy:
+		sortedApps = stateAwareFilter(apps)
+		// Sort by submission time oldest first
+		sort.SliceStable(sortedApps, func(i, j int) bool {
+			l := sortedApps[i]
+			r := sortedApps[j]
+			return l.ApplicationInfo.SubmissionTime < r.ApplicationInfo.SubmissionTime
```

Comment:
The state aware policy looks orthogonal to the choice between Fifo and Fair. In other words: a user could only use state aware policy with the Fifo ordering (but not with fair). I think we should enable this feature for the user.

Reply:
We can add two policies to implement the sorting and still use the state filtering. That would be like adding  a second state aware policy and renaming what we have now.
i.e. StateAwareFIFO and StateAwareFair

Is that what you mean, as that is really simple to implement.